src/core/featuremodel.cpp

```diff
@@ -194,10 +194,21 @@ void FeatureModel::setLinkedFeatureValues()
 {
   beginResetModel();
   mLinkedAttributeIndexes.clear();
+  const bool parentFeatureIsNew = std::numeric_limits<QgsFeatureId>::min() == mLinkedParentFeature.id();
   const auto fieldPairs = mLinkedRelation.fieldPairs();
   for ( QgsRelation::FieldPair fieldPair : fieldPairs )
   {
-    mFeature.setAttribute( mFeature.fieldNameIndex( fieldPair.first ), linkedParentFeature().attribute( fieldPair.second ) );
+    mFeature.setAttribute( fieldPair.first, linkedParentFeature().attribute( fieldPair.second ) );
+    if ( parentFeatureIsNew && mLinkedRelation.referencedLayer() )
+    {
+      if ( mLinkedRelation.referencedLayer()->dataProvider() )
```

Comment:
```suggestion
    if ( parentFeatureIsNew && mLinkedRelation.referencedLayer() && mLinkedRelation.referencedLayer()->dataProvider() )
```
A bit shorter and less nesting.